lib/Bio/EnsEMBL/DataCheck/Checks/CheckFlatProteinTrees.pm

```diff
@@ -0,0 +1,120 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package Bio::EnsEMBL::DataCheck::Checks::CheckFlatProteinTrees;
+
+use warnings;
+use strict;
+
+use Moose;
+use Test::More;
+use Bio::EnsEMBL::DataCheck::Test::DataCheck;
+
+extends 'Bio::EnsEMBL::DataCheck::DbCheck';
+
+use constant {
+  NAME           => 'CheckFlatProteinTrees',
+  DESCRIPTION    => 'Check protein tree integrity ensuring number of leaves with parent node at root < 3',
+  GROUPS         => ['compara', 'compara_protein_trees'],
+  DATACHECK_TYPE => 'critical',
+  DB_TYPES       => ['compara'],
+  TABLES         => ['gene_tree_node', 'gene_tree_root']
+};
+
+sub tests {
+  my ($self) = @_;
+  my $dba = $self->dba;
+  my $helper = $dba->dbc->sql_helper;
+
+  # Count and collect nodes with root_id != 0 and is not a seq_member
+  my $internal_node_sql = qq/
+    SELECT gtn.root_id, COUNT(*) AS internal_nodes 
+      FROM gene_tree_node gtn 
+    WHERE gtn.seq_member_id IS NULL 
+      AND gtn.node_id <> gtn.root_id 
+    GROUP BY gtn.root_id
+  /;
+  my $internal_node = $helper->execute_into_hash(  
+   -SQL => $internal_node_sql
+  );
+
+  # Count and collect seq_members where root_id != 0, parent_id==root_id and count>1
+  my $flat_member_sql = qq/
+    SELECT gtn.root_id, COUNT(*) AS root_members 
+      FROM gene_tree_node gtn 
+    WHERE gtn.seq_member_id IS NOT NULL 
+      AND gtn.parent_id = gtn.root_id 
+    GROUP BY gtn.root_id 
+    HAVING root_members > 1
+  /;
+  my $flat_members = $helper->execute_into_hash(  
+   -SQL => $flat_member_sql,
+  );
+
+  # Count and collect all node memmbers where root_id != 0
+  my $all_members_sql = qq/
+    SELECT root_id, COUNT(DISTINCT root_id) AS root_count 
+      FROM gene_tree_node 
+    GROUP BY root_id
+  /;
+  my $all_members = $helper->execute_into_hash(  
+   -SQL => $all_members_sql
+  );
+
+  # Collect all non-rooted trees
+  my $nonrooted_trees_sql = qq/
+    SELECT root_id, 1
+      FROM gene_tree_root 
+    WHERE tree_type = 'tree' 
+      AND clusterset_id LIKE '%_it_%' 
+      AND clusterset_id NOT LIKE 'pg_it_%'
+  /;
+  my $nonrooted_trees = $helper->execute_into_hash( 
+    -SQL => $nonrooted_trees_sql
+  );
+
+  my @flat_trees;
+  my @flat_trees_w_structure;
+
+  foreach my $root_id ( keys %$flat_members ) {
+    my $count = $flat_members->{$root_id};
+    # Check if $root_id is present in $nonrooted_trees
+    if (exists $internal_node->{$root_id} ) {
+      if ( !exists $nonrooted_trees->{$root_id} ) {
+        push @flat_trees_w_structure, $root_id;
+      } # Check that root has no more than 2 members or 3 if part of non-rooted trees
+      if ( $count > 2 ) {
+        if ( $count == 3 && !exists $nonrooted_trees->{$root_id} ) {
+          push @flat_trees, $root_id;
+        }
+        if ( $count == $all_members->{$root_id} || $count > 3 ) {
```

Comment:
I am confused, isn't that what I am doing? I am checking if the `$count == 3` and is not in the `$nonrooted_trees` then it is flat, so if it has 3 genes and is present as an unrooted tree, then it will not be assigned as a flat tree? Or have I mixed up the logic you meant? :confounded:

Reply:
So `if ( $count == $all_members->{$root_id} && $count > 3 ) {`?